enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,391 @@
+// +openshift:compatibility-gen:level=4
+// +openshift:enable:FeatureSets=TechPreviewNoUpgrade
+type ImagePolicy struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	// +required
+	Spec ImagePolicySpec `json:"spec"`
+	// status contains the observed state of the resource.
+	// +optional
+	Status ImagePolicyStatus `json:"status,omitempty"`
+}
+
+// ImagePolicySpec is the specification of the ImagePolicy CRD.
+type ImagePolicySpec struct {
+	// scopes defines the list of image identities assigned to a policy. Each item refers to an image or repository in a registry implementing the "Docker Registry HTTP API V2".
+	// "scopes" uses one of the following:
+	// - complete image name, either using a tag or digest
+	// - prefixes of individual-image scopes
+	// - a wildcarded expression for matching all subdomains, the wildcard only presents at the beginning, *.example.com is a valid case, but example*.*.com is not.
+	// For more information about the format, see the document
+	// about the docker transport field:
+	// https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md#docker
+	// +kubebuilder:validation:Required
+	// +required
+	// +listType=set
+	Scopes []ImageScope `json:"scopes"`
+	// policy defines the verification policy for the items in the scopes list
+	// +kubebuilder:validation:Required
+	// +required
+	Policy Policy `json:"policy"`
+}
+
+// ImageScope is the item of the scopes list.
+// +kubebuilder:validation:Pattern=`^\*(?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+$|^((?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9])(((?:\.(?:[a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]))+(?::[0-9]+)?)|(:[0-9]+))(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+)?)(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+:([\w][\w.-]{0,127}))?(?:(?:/[a-z0-9]+(?:(?:(?:[._]|__|[-]*)[a-z0-9]+)+)?)+@([A-Za-z][A-Za-z0-9]*(?:[-_+.][A-Za-z][A-Za-z0-9]*)*[:][[:xdigit:]]{32,}))?$`
+type ImageScope string
+
+// Policy defines the verification policy for the items in the scopes list.
+// +kubebuilder:validation:XValidation:rule="(has(self.rootOfTrust) && has(self.rootOfTrust.fulcioCAData)) ? has(self.fulcioSubject) : true",message="fulcioSubject must be set if fulcioCAData is set"
+// +kubebuilder:validation:XValidation:rule="has(self.fulcioSubject) ? has(self.fulcioSubject.oidcIssuer) && has(self.fulcioSubject.signedEmail): true",message="oidcIssuer and signedEmail must be set if fulcioSubject is set"
+// +kubebuilder:validation:XValidation:rule="has(self.signedIdentity) ? has(self.signedIdentity.matchPolicy): true",message="matchPolicy must be set if signedIdentity is set"
```

Comment:
Yet the field is defined as optional. Is that intentional?

(The concern about duplicating the validation applies similarly, though even less urgently because `signedIdentity` is a bit less likely to need expansion.)

Reply:
mark the `matchPolicy` required  . 

Dropped these validations from the API defination. I just remembered that we can move these dependency validation to https://github.com/openshift/kubernetes/tree/master/openshift-kube-apiserver/admission/customresourcevalidation.

Reply:
I’m afraid I have no idea what is the difference and tradeoffs between the two approaches.

Reply:
> My (not very important) concern here is that it seems possible that we will add more fields / alternatives to FulcioSubject, and in that case it seems easier to maintain if the validation of the required fields / allowed field combinations only exists in one place, instead of two like it is now.

I thought the https://github.com/openshift/kubernetes/tree/master/openshift-kube-apiserver/admission/customresourcevalidation give us one place to put these validations.  But openshift/api provides a way to test these validations by directly reading the manifest yaml.